torch/nn/_functions/rnn.py

```diff
@@ -10,26 +10,29 @@
 
 
 def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.relu(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.relu(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.tanh(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.tanh(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     if input.is_cuda:
-        igates = F.linear(input, w_ih)
+        igates = input.expand(4, input.size(0), input.size(1)).transpose(0, 1) if w_ih is None else F.linear(input,
```

Comment:
so with this implementation you would have something like: lstm_skip(128) >>= linear(128, 128) >>= lstm_skip(128). this really should be lstm_skip(128) >>= linear(128, 512) >>= lstm_skip(128) in order to be able to get the same results as normal stacked lstms. looking deeper, the first version would be equivalent to... 
let E = [eye(128), eye(128), eye(128), eye(128)] in
LSTM(E(W * LSTM(E(x)))).

Whereas with the raw multilayer cudnn implementation it would be LSTM(W * LSTM(E(x))), which is still wrong, but less wrong.